### ISSUE TITLE
tensorflow embedding intent classifier GPU optimization issue

### ISSUE THREAD

**Issue by princerk** [RasaHQ/rasa_nlu#1324]
**Rasa NLU version**: 0.13.1

**Operating system** (windows, osx, ...): ubuntu

**Content of model configuration file**:
```yml
language: "en"

pipeline:
- name: "intent_featurizer_count_vectors"
- name: "intent_classifier_tensorflow_embedding"
  "epochs": 3000
```

**Issue**: Training time on GPU and CPU are almost same. I tested this with Nvidia P100 GPU on google club services.

**Comment by amn41** [comment]
@princerk thanks for testing this out, but I'm afraid that without more information this is impossible to comment on. 
1. how much data do you have?
2. how long does training take?
3. what kind of speedup were you expecting? running on a GPU doesn't guarantee things will always be faster, there is overhead in moving computations to the GPU
4. are you sure tensorflow is able to see and run on the GPU?

**Comment by lisongshan** [comment]
Hi, @amn41 Does Rasa core support tensorflow-gpu?

**Comment by princerk** [comment]
Hi @amn41, thanks for quick reply
I tested on with the example data given the this repo 
              rasa_nlu/data/examples/rasa/demo-rasa.json
Currently it has only 42 examples with 4 intents, to increase the data size I copied the same training data multiple times. taking it to 170K examples, just to see how does it affect the training time on GPU vs CPU. Initially for small data size I waited to complete the training but for large data size I only saw the it/s. I am sure it is using the GPU as you can see in the following log message.





2018-08-17 15:51:16 INFO     rasa_nlu.training_data.loading  - Training data format of demo-rasa.json is rasa_nlu
2018-08-17 15:51:16 INFO     rasa_nlu.training_data.training_data  - Training data stats: 
	- intent examples: 42 (4 distinct intents)
	- Found intents: 'restaurant_search', 'greet', 'affirm', 'goodbye'
	- entity examples: 11 (2 distinct entities)
	- found entities: 'cuisine', 'location'

2018-08-17 15:51:16 INFO     rasa_nlu.model  - Starting to train component intent_featurizer_count_vectors
2018-08-17 15:51:16 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 15:51:16 INFO     rasa_nlu.model  - Starting to train component intent_classifier_tensorflow_embedding
2018-08-17 15:51:21.075255: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-17 15:51:21.263911: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-08-17 15:51:21.264378: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: Tesla P100-PCIE-16GB major: 6 minor: 0 memoryClockRate(GHz): 1.3285
pciBusID: 0000:00:04.0
totalMemory: 15.90GiB freeMemory: 15.51GiB
2018-08-17 15:51:21.264414: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-08-17 15:51:22.906366: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-17 15:51:22.906455: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-08-17 15:51:22.906466: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-08-17 15:51:22.907796: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 15031 MB memory) -> physical GPU (device: 0, name: Tesla P100-PCIE-16GB, pci bus id: 0000:00:04.0, compute capability: 6.0)
2018-08-17 15:51:23 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Accuracy is updated every 10 epochs
Epochs: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3000/3000 [00:17<00:00, 173.59it/s, loss=0.004, acc=1.000]
2018-08-17 15:51:40 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Finished training embedding policy, loss=0.004, train accuracy=1.000
2018-08-17 15:51:40 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 15:51:40 INFO     rasa_nlu.model  - Successfully saved model into '/home/r0k00d7/rasa-gpu-test/projects/default/model_20180817-155140'
2018-08-17 15:51:40 INFO     __main__  - Finished training

**Comment by amn41** [comment]
thanks @princerk ! I wouldn't expect any speed difference because this is a tiny amount of data, so all the wall-clock time is spent initializing. If you have a dataset with a few thousand training examples I would expect you to start seeing differences

**Comment by princerk** [comment]
@amn41 For small data size CPU out performed GPU. I thought maybe because the file transfer to the GPU it taking more time. 
Find below the log for 2688 examples on GPU vs CPU, If you want I can provide the log with more examples.

**GPU log:**
2018-08-17 16:11:30 INFO     rasa_nlu.training_data.loading  - Training data format of demo-rasa.json is rasa_nlu
2018-08-17 16:11:30 INFO     rasa_nlu.training_data.training_data  - Training data stats: 
	- intent examples: 2688 (4 distinct intents)
	- Found intents: 'greet', 'restaurant_search', 'affirm', 'goodbye'
	- entity examples: 704 (2 distinct entities)
	- found entities: 'cuisine', 'location'

2018-08-17 16:11:30 INFO     rasa_nlu.model  - Starting to train component intent_featurizer_count_vectors
2018-08-17 16:11:30 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:11:30 INFO     rasa_nlu.model  - Starting to train component intent_classifier_tensorflow_embedding
2018-08-17 16:11:33.574997: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-17 16:11:33.695800: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-08-17 16:11:33.696252: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: Tesla P100-PCIE-16GB major: 6 minor: 0 memoryClockRate(GHz): 1.3285
pciBusID: 0000:00:04.0
totalMemory: 15.90GiB freeMemory: 15.51GiB
2018-08-17 16:11:33.696286: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-08-17 16:11:34.009790: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-17 16:11:34.009876: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-08-17 16:11:34.009886: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-08-17 16:11:34.010244: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 15031 MB memory) -> physical GPU (device: 0, name: Tesla P100-PCIE-16GB, pci bus id: 0000:00:04.0, compute capability: 6.0)
2018-08-17 16:11:34 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Accuracy is updated every 10 epochs
Epochs: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:45<00:00,  6.64it/s, acc=1.000, loss=0.006]
2018-08-17 16:12:19 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Finished training embedding policy, loss=0.006, train accuracy=1.000
2018-08-17 16:12:19 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:12:19 INFO     rasa_nlu.model  - Successfully saved model into '/home/r0k00d7/rasa-gpu-test/projects/default/model_20180817-161219'
2018-08-17 16:12:19 INFO     __main__  - Finished training


**CPU log:**
2018-08-17 16:13:12 INFO     rasa_nlu.training_data.loading  - Training data format of demo-rasa.json is rasa_nlu
2018-08-17 16:13:12 INFO     rasa_nlu.training_data.training_data  - Training data stats: 
	- intent examples: 2688 (4 distinct intents)
	- Found intents: 'restaurant_search', 'greet', 'affirm', 'goodbye'
	- entity examples: 704 (2 distinct entities)
	- found entities: 'cuisine', 'location'

2018-08-17 16:13:13 INFO     rasa_nlu.model  - Starting to train component intent_featurizer_count_vectors
2018-08-17 16:13:13 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:13:13 INFO     rasa_nlu.model  - Starting to train component intent_classifier_tensorflow_embedding
2018-08-17 16:13:16.338049: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-17 16:13:16 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Accuracy is updated every 10 epochs
Epochs: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [00:42<00:00,  7.12it/s, acc=1.000, loss=0.002]
2018-08-17 16:13:58 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Finished training embedding policy, loss=0.002, train accuracy=1.000
2018-08-17 16:13:58 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:13:58 INFO     rasa_nlu.model  - Successfully saved model into '/home/r0k00d7/rasa-gpu-test/projects/default/model_20180817-161358'
2018-08-17 16:13:58 INFO     __main__  - Finished training

**Comment by princerk** [comment]
**GPU log:**
2018-08-17 16:35:08 INFO     rasa_nlu.training_data.loading  - Training data format of demo-rasa.json is rasa_nlu
2018-08-17 16:35:10 INFO     rasa_nlu.training_data.training_data  - Training data stats: 
	- intent examples: 43008 (4 distinct intents)
	- Found intents: 'goodbye', 'restaurant_search', 'greet', 'affirm'
	- entity examples: 11264 (2 distinct entities)
	- found entities: 'cuisine', 'location'

2018-08-17 16:35:12 INFO     rasa_nlu.model  - Starting to train component intent_featurizer_count_vectors
2018-08-17 16:35:12 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:35:12 INFO     rasa_nlu.model  - Starting to train component intent_classifier_tensorflow_embedding
2018-08-17 16:35:16.400323: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-17 16:35:16.510552: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-08-17 16:35:16.511031: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties: 
name: Tesla P100-PCIE-16GB major: 6 minor: 0 memoryClockRate(GHz): 1.3285
pciBusID: 0000:00:04.0
totalMemory: 15.90GiB freeMemory: 15.53GiB
2018-08-17 16:35:16.511065: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-08-17 16:35:16.830469: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-17 16:35:16.830534: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 
2018-08-17 16:35:16.830558: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N 
2018-08-17 16:35:16.830866: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 15049 MB memory) -> physical GPU (device: 0, name: Tesla P100-PCIE-16GB, pci bus id: 0000:00:04.0, compute capability: 6.0)
2018-08-17 16:35:17 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Accuracy is updated every 10 epochs
Epochs: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [12:37<00:00,  2.52s/it, acc=1.000, loss=0.002]
2018-08-17 16:47:54 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Finished training embedding policy, loss=0.002, train accuracy=1.000
2018-08-17 16:47:54 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:47:55 INFO     rasa_nlu.model  - Successfully saved model into '/home/r0k00d7/rasa-gpu-test/projects/default/model_20180817-164754'
2018-08-17 16:47:55 INFO     __main__  - Finished training


**CPU log:**
2018-08-17 16:19:56 INFO     rasa_nlu.training_data.loading  - Training data format of demo-rasa.json is rasa_nlu
2018-08-17 16:19:58 INFO     rasa_nlu.training_data.training_data  - Training data stats: 
	- intent examples: 43008 (4 distinct intents)
	- Found intents: 'affirm', 'greet', 'restaurant_search', 'goodbye'
	- entity examples: 11264 (2 distinct entities)
	- found entities: 'cuisine', 'location'

2018-08-17 16:19:59 INFO     rasa_nlu.model  - Starting to train component intent_featurizer_count_vectors
2018-08-17 16:20:00 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:20:00 INFO     rasa_nlu.model  - Starting to train component intent_classifier_tensorflow_embedding
2018-08-17 16:20:03.936169: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-17 16:20:03 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Accuracy is updated every 10 epochs
Epochs: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 300/300 [12:11<00:00,  2.44s/it, loss=0.002, acc=1.000]2018-08-17 16:32:15 INFO     rasa_nlu.classifiers.embedding_intent_classifier  - Finished training embedding policy, loss=0.002, train accuracy=1.000
2018-08-17 16:32:15 INFO     rasa_nlu.model  - Finished training component.
2018-08-17 16:32:15 INFO     rasa_nlu.model  - Successfully saved model into '/home/r0k00d7/rasa-gpu-test/projects/default/model_20180817-163215'
2018-08-17 16:32:15 INFO     __main__  - Finished training

**Comment by princerk** [comment]
I understand there are many factors which determines whether the training on GPU would take less time than CPU, one of the factor is the computation that is being done, If most the time goes in reading the data from memory then it is not expected to gain performance from the GPU, to test this out I am increasing the embedding dimensions and the number of layers, and will see how it goes, any other suggestion would be really helpful. because I am getting good classification result with the default number of layers.

**Comment by princerk** [comment]
@lisongshan rasa_core supports training on GPU. I am not sure about the improvement in training time,I tried with get started example given in rasa_core docs.
